File: doc-examples/mongo-example-java/src/main/java/example/Sale.java

package example;

import io.micronaut.data.annotation.GeneratedValue;
import io.micronaut.data.annotation.Id;
import io.micronaut.data.annotation.MappedEntity;
import io.micronaut.data.annotation.MappedProperty;
import io.micronaut.data.annotation.Relation;
import org.bson.types.ObjectId;

@MappedEntity
public class Sale {

    @Relation(Relation.Kind.MANY_TO_ONE)
    private final Product product;
    @MappedProperty(converter = QuantityAttributeConverter.class)
    private final Quantity quantity;

    @Id
    @GeneratedValue
    private ObjectId id;

    public Sale(Product product, Quantity quantity) {
        this.product = product;
        this.quantity = quantity;
    }

    public Product getProduct() {
        return product;
    }

    public Quantity getQuantity() {
        return quantity;
    }

    public ObjectId getId() {
        return id;
    }

    public void setId(ObjectId id) {
        this.id = id;
    }
}

File: doc-examples/mongo-example-python/src/main/python/example/Sale.py
from dataclasses import dataclass
from typing import Annotated

from micronaut.data.annotation import GeneratedValue, Id, MappedEntity, MappedProperty, Relation
from org.bson.types import ObjectId

from example.Product import Product
from example.Quantity import Quantity
from example.QuantityAttributeConverter import QuantityAttributeConverter


@MappedEntity
@dataclass
class Sale:
    product: Annotated[Product, Relation("MANY_TO_ONE")]
    quantity: Annotated[Quantity, MappedProperty(converter=QuantityAttributeConverter)]
    id: Annotated[ObjectId | None, Id, GeneratedValue] = None
